In [1]:
import pandas as pd
import numpy as np
import json
from datetime import datetime

In [2]:
df_lst = []
for i in range(56):
    df_lst.append( pd.read_csv('./data/Kickstarter{:03d}.csv'.format(i)) )

In [3]:
df = pd.concat(df_lst, ignore_index=True)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 209222 entries, 0 to 209221
Data columns (total 37 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   backers_count             209222 non-null  int64  
 1   blurb                     209214 non-null  object 
 2   category                  209222 non-null  object 
 3   converted_pledged_amount  209222 non-null  int64  
 4   country                   209222 non-null  object 
 5   created_at                209222 non-null  int64  
 6   creator                   209222 non-null  object 
 7   currency                  209222 non-null  object 
 8   currency_symbol           209222 non-null  object 
 9   currency_trailing_code    209222 non-null  bool   
 10  current_currency          209222 non-null  object 
 11  deadline                  209222 non-null  int64  
 12  disable_communication     209222 non-null  bool   
 13  friends                   300 non-null     o

In [5]:
df.head()

backers_count                                              blurb  \
0             21  2006 was almost 7 years ago.... Can you believ...   
1             97  An adorable fantasy enamel pin series of princ...   
2             88  Helping a community come together to set the s...   
3            193  Every revolution starts from the bottom and we...   
4             20  Learn to build 10+ Applications in this comple...   

                                            category  \
0  {"id":43,"name":"Rock","slug":"music/rock","po...   
1  {"id":54,"name":"Mixed Media","slug":"art/mixe...   
2  {"id":280,"name":"Photobooks","slug":"photogra...   
3  {"id":266,"name":"Footwear","slug":"fashion/fo...   
4  {"id":51,"name":"Software","slug":"technology/...   

   converted_pledged_amount country  created_at  \
0                       802      US  1387659690   
1                      2259      US  1549659768   
2                     29638      US  1477242384   
3                     49158      IT  1540369920   
4                       549      US  1425706517   

                                             creator currency currency_symbol  \
0  {"id":1495925645,"name":"Daniel","is_registere...      USD               $   
1  {"id":1175589980,"name":"Katherine","slug":"fr...      USD               $   
2  {"id":1196856269,"name":"MelissaThomas","is_re...      USD               $   
3  {"id":1569700626,"name":"WAO","slug":"wearewao...      EUR               €   
4  {"id":1870845385,"name":"Kalpit Jain","is_regi...      USD               $   

   currency_trailing_code  ...  \
0                    True  ...   
1                    True  ...   
2                    True  ...   
3                   False  ...   
4                    True  ...   

                                                slug  \
0                              new-final-round-album   
1                    princess-pals-enamel-pin-series   
2  their-life-through-their-lens-the-amish-and-me...   
3                           wao-the-eco-effect-shoes   
4                     apple-watch-development-course   

                                          source_url  spotlight staff_pick  \
0  https://www.kickstarter.com/discover/categorie...       True      False   
1  https://www.kickstarter.com/discover/categorie...       True      False   
2  https://www.kickstarter.com/discover/categorie...       True       True   
3  https://www.kickstarter.com/discover/categorie...       True      False   
4  https://www.kickstarter.com/discover/categorie...      False      False   

        state  state_changed_at  static_usd_rate  \
0  successful        1391899046         1.000000   
1  successful        1551801611         1.000000   
2  successful        1480607932         1.000000   
3  successful        1544309940         1.136525   
4      failed        1428511019         1.000000   

                                                urls  usd_pledged  \
0  {"web":{"project":"https://www.kickstarter.com...    802.00000   
1  {"web":{"project":"https://www.kickstarter.com...   2259.00000   
2  {"web":{"project":"https://www.kickstarter.com...  29638.00000   
3  {"web":{"project":"https://www.kickstarter.com...  49075.15252   
4  {"web":{"project":"https://www.kickstarter.com...    549.00000   

        usd_type  
0  international  
1  international  
2  international  
3  international  
4       domestic  

[5 rows x 37 columns]

# Feature overview

| Column | Description | Dtype | Null | EDA status |
| - | - | - | - | - |
| `backers_count` | The number of participants of the kickstarter project | | | Done |
| `blurb` | Short description of the project by the project initiator. Max 135 characters. | | | Done | 
| `category` | Dictionary object. \n Contains `id, name, slug, position, parent_id, color, urls` | | | Open |
| `converted_pledged_amount` | Open | | | Open |
| `country` | Country of of Kickstarter account (?) | | | To be confirmed |
| `created_at`| Creation date of the project. UNIX time format | | | Done |
| `creator`| Dictionary of user details `id, name, is_registered, chosen_currency, ... | | | Open to list all dictionary entries|
| `currency`| Currency (3-digit code) by which the project is founded | | | Done |
| `currency_symbol`| Symbol related to `currency` feature. | | | Done |
| `currency_trailing_code`| | Bool | | Open |
| `current_currency`| Open | | | Open |
| `deadline`| Deadline of the project. UNIX time format | | | |
| `disable_communication`| Is the communication with the creator allowed? | Bool | | Done |
| `friends`| 300 entries with empty list object | | | See '300 question' |
| `fx_rate`| Open | | | Open. Currency topic. |
| `goal`| Target amount | | | By how much deviate the projects from the target? (successful vs. failed) |
| `id`| Intern kickstarter id | | | See '182-anomalia' |
| `is_backing`| 300 instances of 'False' | | | See '300 question' |
| `is_starrable`| Open | Bool | None | Open |
| `is_starred`| Open | | 298 False, 2 True | See '300 question' |
| `launched_at`| Date of publication on kickstarter | int64 | None | Done |
| `location`| Dictionary of location details | String | 208996 filled | Open |
| `name`| Project name. Limited to 60 chars. | String | 209222 filled | Open |
| `permissions`| 300 entries with empty list object ||| See '300 question' |
| `photo`| Dictionary of photo url, ...  | String | | See '182-anomalia' |
| `pledged`|||||
| `profile`|||||
| `slug`|||||
| `source_url`|||||
| `spotlight`|||||
| `staff_pick`|||||
| `state`|||||
| `state_changed_at`|||||
| `static_usd_rate`|||||
| `urls`|||||
| `usd_pledged`|||||
| `usd_type` |||||

# Data cleaning

## JSON objects

In [6]:
def get_json_cols(data):
    cols_object = df.select_dtypes(include='object').columns
    return [x for x in cols_object if any(data[x].astype(str).str.contains('{\"'))]

In [7]:
def create_dicts_from_json(data, cols_json):
    data_dicts = pd.DataFrame()
    for col in cols_json:
        print('Parsing json in: '+col)
        c = []
        for i, val in data[col].items():
            try:
                c.append(json.loads(val))
            except:
                c.append(dict())
        data_dicts[col] = pd.Series(np.array(c))
    return data_dicts

In [8]:
def create_cols_from_dicts(data_dicts):
    data_expanded = []
    for col in data_dicts.columns:
        print('Expanding: '+col)
        data_expanded.append(pd.json_normalize(data_dicts[col]).add_prefix(col+'_'))
    return pd.concat(data_expanded, axis=1)

In [9]:
def save_dataframe(data, file_name=None):
    if file_name:
        e = file_name
    else:
        t = datetime.now().strftime("%Y-%m-%d_%H%M%S")
        e = f"./data_frame_{t}.pickle"
    print('Saving: '+e)
    data.to_pickle(e)

In [10]:
def expand_json(data, file_name=None, save=True):
    #from datetime import datetime
    cols_json = get_json_cols(data)
    print('---------- Parsing json ------------')
    data_dicts = create_dicts_from_json(data, cols_json)
    print('---------- Expanding dictionaries --------')
    data_expanded = create_cols_from_dicts(data_dicts)
    print('---------- Merge to final data frame ------')
    data_merged = pd.concat([data.drop(cols_json, axis=1), data_expanded], axis=1)
    if save:
        print('---------- Saving to pickle ------')
        save_dataframe(data_merged)
    return data_merged

# Full raw data frame

In [11]:
df_new = expand_json(df, file_name='./data_frame_raw_2021-03-12.pickle')

---------- Parsing json ------------
Parsing json in: category
Parsing json in: creator
Parsing json in: location
Parsing json in: photo
Parsing json in: profile
Parsing json in: urls
---------- Expanding dictionaries --------
Expanding: category
Expanding: creator
Expanding: location
Expanding: photo
Expanding: profile
Expanding: urls
---------- Merge to final data frame ------
---------- Saving to pickle ------
Saving: ./data_frame_2021-03-12_075227.pickle


In [12]:
df_new.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 209222 entries, 0 to 209221
Data columns (total 96 columns):
 #   Column                                                        Non-Null Count   Dtype  
---  ------                                                        --------------   -----  
 0   backers_count                                                 209222 non-null  int64  
 1   blurb                                                         209214 non-null  object 
 2   converted_pledged_amount                                      209222 non-null  int64  
 3   country                                                       209222 non-null  object 
 4   created_at                                                    209222 non-null  int64  
 5   currency                                                      209222 non-null  object 
 6   currency_symbol                                               209222 non-null  object 
 7   currency_trailing_code                                  